[Quant] Separate FBGEMM/QNNPACK BackendConfigs

andrewor14 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #83566

Summary: Previously we use a single BackendConfig
(get_native_backend_config) for both the FBGEMM and QNNPACK
backends. However, these two backends have subtle differences
in terms of their requirements that cannot be satisfied using
a single BackendConfig. Therefore, this commit is the first step
torwards decoupling the two backends. The real change in
functionality will come in a future commit after DTypeConfig
supports quant_min/quant_max and scale_min/scale_max. Existing
uses of `get_native_backend_config` should not be affected.

Public facing changes:
```
from torch.ao.quantization.backend_config import (
    get_fbgemm_backend_config,
    get_qnnpack_backend_config,
)
fbgemm_backend_config = get_fbgemm_backend_config()
qnnpack_backend_config = get_qnnpack_backend_config()
```

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168

Subscribers: jerryzh168

andrewor14 committed:
[Quant] Separate FBGEMM/QNNPACK BackendConfigs

Summary: Previously we use a single BackendConfig
(get_native_backend_config) for both the FBGEMM and QNNPACK
backends. However, these two backends have subtle differences
in terms of their requirements that cannot be satisfied using
a single BackendConfig. Therefore, this commit is the first step
torwards decoupling the two backends. The real change in
functionality will come in a future commit after DTypeConfig
supports quant_min/quant_max and scale_min/scale_max. Existing
uses of `get_native_backend_config` should not be affected.

Public facing changes:
```
from torch.ao.quantization.backend_config import (
    get_fbgemm_backend_config,
    get_qnnpack_backend_config,
)
fbgemm_backend_config = get_fbgemm_backend_config()
qnnpack_backend_config = get_qnnpack_backend_config()
```

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168

Subscribers: jerryzh168

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant] Separate FBGEMM/QNNPACK BackendConfigs"


Summary: Previously we use a single BackendConfig
(get_native_backend_config) for both the FBGEMM and QNNPACK
backends. However, these two backends have subtle differences
in terms of their requirements that cannot be satisfied using
a single BackendConfig. Therefore, this commit is the first step
torwards decoupling the two backends. The real change in
functionality will come in a future commit after DTypeConfig
supports quant_min/quant_max and scale_min/scale_max. Existing
uses of `get_native_backend_config` should not be affected.

Public facing changes:
```
from torch.ao.quantization.backend_config import (
    get_fbgemm_backend_config,
    get_qnnpack_backend_config,
)
fbgemm_backend_config = get_fbgemm_backend_config()
qnnpack_backend_config = get_qnnpack_backend_config()
```

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168

Subscribers: jerryzh168

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/backend_config/_common_operator_config_utils.py:
```diff
@@ -10,6 +10,7 @@
 from typing import List
 from .observation_type import ObservationType
 from .backend_config import BackendPatternConfig, DTypeConfig
+from ..fake_quantize import FixedQParamsFakeQuantize
```
do we still need to use this? or can we just use normal FakeQuantize?

jerryzh168 replied:
sure, sounds good

jerryzh168 commented on torch/ao/quantization/backend_config/fbgemm.py:
```diff
@@ -0,0 +1,102 @@
+        fbgemm_weight_only_quint8_dtype_config,
+        fbgemm_weight_only_quint4x2_dtype_config,
+    ]
+    return _get_native_backend_config_helper(
```
I feel it might be better to do 
```
BackendConfig("_native_and_fp16") \
        .set_backend_pattern_configs(_DEFAULT_OP_INT8_CONFIGS) \
        .set_backend_pattern_configs(_get_linear_configs(linear_dtype_configs)) \
        .set_backend_pattern_configs(_get_conv_configs(conv_dtype_configs)) \
        .set_backend_pattern_configs(_get_binary_op_configs(binary_op_dtype_configs)) \
        .set_backend_pattern_configs(_get_fixed_qparams_op_configs(fixed_qparams_op_dtype_configs)) \
        .set_backend_pattern_config(_CAT_CONFIG) \
        .set_backend_pattern_configs(_get_bn_configs()) \
        .set_backend_pattern_configs(_get_share_qparams_op_configs(share_qparams_op_dtype_configs)) \
        .set_backend_pattern_configs(_get_rnn_op_configs()) \
        .set_backend_pattern_configs(_get_embedding_op_configs())
```
this will be easier to extend if in the future we need to add new pattern configs

andrewor14 replied:
Ok. It's just that we will be duplicating those lines 3 times (1 here, 1 for fbgemm, and 1 for qnnpack)

jerryzh168 replied:
yeah I think that is fine

jerryzh168 reviewed (approved):
lgtm, thanks a lot! had a few inline comments

andrewor14 committed:
Update on "[Quant] Separate FBGEMM/QNNPACK BackendConfigs"


Summary: Previously we use a single BackendConfig
(get_native_backend_config) for both the FBGEMM and QNNPACK
backends. However, these two backends have subtle differences
in terms of their requirements that cannot be satisfied using
a single BackendConfig. Therefore, this commit is the first step
torwards decoupling the two backends. The real change in
functionality will come in a future commit after DTypeConfig
supports quant_min/quant_max and scale_min/scale_max. Existing
uses of `get_native_backend_config` should not be affected.

Public facing changes:
```
from torch.ao.quantization.backend_config import (
    get_fbgemm_backend_config,
    get_qnnpack_backend_config,
)
fbgemm_backend_config = get_fbgemm_backend_config()
qnnpack_backend_config = get_qnnpack_backend_config()
```

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168

Subscribers: jerryzh168

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant] Separate FBGEMM/QNNPACK BackendConfigs"


Summary: Previously we use a single BackendConfig
(get_native_backend_config) for both the FBGEMM and QNNPACK
backends. However, these two backends have subtle differences
in terms of their requirements that cannot be satisfied using
a single BackendConfig. Therefore, this commit is the first step
torwards decoupling the two backends. The real change in
functionality will come in a future commit after DTypeConfig
supports quant_min/quant_max and scale_min/scale_max. Existing
uses of `get_native_backend_config` should not be affected.

Public facing changes:
```
from torch.ao.quantization.backend_config import (
    get_fbgemm_backend_config,
    get_qnnpack_backend_config,
)
fbgemm_backend_config = get_fbgemm_backend_config()
qnnpack_backend_config = get_qnnpack_backend_config()
```

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168

Subscribers: jerryzh168

[ghstack-poisoned]

andrewor14 commented:
@pytorchmergebot merge
